content/posts/2024-08-31-the-1001-thresholds-in-llvm.md

```diff
@@ -0,0 +1,66 @@
+
+<div style="max-width:600px; margin:0 auto;">
+  <img src="/img/the-1001-thresholds-in-llvm-2024-08-31-figure1.png"><br />
+</div>
+
+<div style="max-width:600px; margin:0 auto;">
+  <img src="/img/the-1001-thresholds-in-llvm-2024-08-31-figure2.png"><br />
+</div>
+
+# Results
+
+The per file study proves that there is no one single magic knob value and the optimum, with regards to compile time or code size, depends on the file that is currently being compiled. For instance here we can see that different knob values (for the knob licm-mssa-optimization-cap) give good cumulative compile time improvements for different files. In detail, most files benefit from a knob value of 300 while 60 is the best knob value for the second most files.  
+
+<div style="max-width:600px; margin:0 auto;">
+  <img src="/img/the-1001-thresholds-in-llvm-2024-08-31-figure3.png"><br />
+</div>
+
+We further show that the presence of an oracle that can tell the best knob value for each file can significantly improve the cumulative compile time. 
+
+<div style="max-width:600px; margin:0 auto;">
+  <img src="/img/the-1001-thresholds-in-llvm-2024-08-31-figure4.png"><br />
+</div>
+
+In this project, we explored various thresholds in LLVM—specifically, 93 thresholds using the Clang matcher—and observed that these thresholds are largely file-specific. This indicates that there is no universally optimal value, or even a set of values, that can be applied across different scenarios. Instead, what is needed is an adaptive mechanism within LLVM, an oracle, that can dynamically determine the appropriate threshold values during compilation.
+
+We also experimented with varying thresholds cumulatively by leveraging file-specific information through an LLVM pass. However, after discussions with the mentors, this approach was set aside due to the significant changes it would necessitate across other parts of the LLVM codebase.
+
+As a result, we have not yet categorized different thresholds, such as identifying optimal threshold values for specific file types (e.g., I/O-intensive files). Nonetheless, we provide substantial [statistical evidence](https://sh0g0-1758.github.io/GSOC-X-LLVM/) in the form of LLVM statistics, bitcode size, and compile-time graphs, as well as histograms that examine these variations on a per-file basis. Additionally, a correlation table between knob values and performance metrics further illustrates the significant impact this study could have on improving LLVM's overall performance.
```

Comment:
I do not see any statistical evidence on the link provided in the common sense of this word. Just bunch of numbers without any conclusion, reasoning and analysis. How these results could be used? In some sense they are trivial and more or less well-known to the community.

Reply:
@asl , I understand what you are saying but we never got to that part, it was mostly research to develop some experimental setup to collect these results.

cc: @jdoerfert @jhueckelheim